File: frontend/src/api/audit.ts
import { apiFetch } from "./client";

export type AuditSettingsOut = {
  enabled: boolean;
  retention_days: number;
};

export type AuditEventOut = {
  id: string;
  entity_type: string;
  entity_id: string | null;
  action: string;
  actor_user_id: string | null;
  actor_name: string | null;
  details_json: string | null;
  created_at: string;
};

export async function getAuditSettings(): Promise<AuditSettingsOut> {
  return apiFetch<AuditSettingsOut>("/api/v1/audit/settings");
}

export async function patchAuditSettings(body: {
  enabled?: boolean;
  retention_days?: number;
}): Promise<AuditSettingsOut> {
  return apiFetch<AuditSettingsOut>("/api/v1/audit/settings", {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function listAuditEvents(params?: {
  limit?: number;
  offset?: number;
  entity_type?: string;
  action?: string;
  q?: string;
  actor_user_id?: string;
  actor_q?: string;
  system_only?: boolean;
}): Promise<AuditEventOut[]> {
  const sp = new URLSearchParams();
  if (typeof params?.limit === "number") sp.set("limit", String(params.limit));
  if (typeof params?.offset === "number") sp.set("offset", String(params.offset));
  if (params?.entity_type?.trim()) sp.set("entity_type", params.entity_type.trim());
  if (params?.action?.trim()) sp.set("action", params.action.trim());
  if (params?.q?.trim()) sp.set("q", params.q.trim());
  if (params?.actor_user_id?.trim()) sp.set("actor_user_id", params.actor_user_id.trim());
  if (params?.actor_q?.trim()) sp.set("actor_q", params.actor_q.trim());
  if (params?.system_only) sp.set("system_only", "true");
  const qs = sp.toString();
  return apiFetch<AuditEventOut[]>(`/api/v1/audit/events${qs ? `?${qs}` : ""}`);
}

File: backend/app/routers/audit.py
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.deps import require_permission
from app.models import User
from app.permissions import SYSTEMS_MANAGE
from app.schemas.audit import AuditEventOut, AuditSettingsOut, AuditSettingsPatch
from app.services.audit import (
    get_audit_enabled,
    get_audit_retention_days,
    list_audit_events,
    set_audit_settings,
)

router = APIRouter(prefix="/audit", tags=["audit"])


@router.get("/settings", response_model=AuditSettingsOut)
async def get_settings(
    session: Annotated[AsyncSession, Depends(get_db)],
    _: Annotated[User, Depends(require_permission(SYSTEMS_MANAGE))],
) -> AuditSettingsOut:
    return AuditSettingsOut(
        enabled=await get_audit_enabled(session),
        retention_days=await get_audit_retention_days(session),
    )


@router.patch("/settings", response_model=AuditSettingsOut)
async def patch_settings(
    body: AuditSettingsPatch,
    session: Annotated[AsyncSession, Depends(get_db)],
    _: Annotated[User, Depends(require_permission(SYSTEMS_MANAGE))],
) -> AuditSettingsOut:
    enabled, retention_days = await set_audit_settings(
        session, enabled=body.enabled, retention_days=body.retention_days
    )
    return AuditSettingsOut(enabled=enabled, retention_days=retention_days)


@router.get("/events", response_model=list[AuditEventOut])
async def get_events(
    session: Annotated[AsyncSession, Depends(get_db)],
    _: Annotated[User, Depends(require_permission(SYSTEMS_MANAGE))],
    limit: Annotated[int, Query(ge=1, le=500)] = 100,
    offset: Annotated[int, Query(ge=0)] = 0,
    entity_type: str | None = None,
    action: str | None = None,
    q: str | None = None,
) -> list[AuditEventOut]:
    rows = await list_audit_events(
        session,
        limit=limit,
        offset=offset,
        entity_type=entity_type,
        action=action,
        q=q,
    )
    user_ids = {r.actor_user_id for r in rows if r.actor_user_id is not None}
    names_by_id: dict[uuid.UUID, str] = {}
    if user_ids:
        users = await session.execute(select(User.id, User.full_name).where(User.id.in_(user_ids)))
        names_by_id = {uid: name for uid, name in users.all()}
    return [
        AuditEventOut(
            id=r.id,
            entity_type=r.entity_type,
            entity_id=r.entity_id,
            action=r.action,
            actor_user_id=r.actor_user_id,
            actor_name=names_by_id.get(r.actor_user_id) if r.actor_user_id else None,
            details_json=r.details_json,
            created_at=r.created_at,
        )
        for r in rows
    ]
